src/compiler/checker.ts

```diff
@@ -23326,13 +23326,39 @@ namespace ts {
+                                            inferFromTypes(sliceTupleType(source, startLength, endLength + sourceArity - impliedArity), elementTypes[startLength]);
+                                            inferFromTypes(sliceTupleType(source, startLength + impliedArity, endLength), elementTypes[startLength + 1]);
+                                        }
+                                    }
+                                    else if (elementFlags[startLength] & ElementFlags.Variadic && elementFlags[startLength + 1] & ElementFlags.Rest) {
+                                        // Middle of target is [...T, ...rest] and source is tuple type
+                                        // if T is constrained by a fixed-size tuple we might be able to use its arity to infer T
+                                        const param = getInferenceInfoForType(elementTypes[startLength])?.typeParameter;
+                                        const constraint = param && getBaseConstraintOfType(param);
+                                        if (constraint && isTupleType(constraint) && !constraint.target.hasRestElement) {
+                                            const impliedArity = constraint.target.fixedLength;
+                                            inferFromTypes(sliceTupleType(source, startLength, sourceArity - (startLength + impliedArity)), elementTypes[startLength]);
+                                        }
+                                    }
+                                    else if (elementFlags[startLength] & ElementFlags.Rest && elementFlags[startLength + 1] & ElementFlags.Variadic && isTupleType(source)) {
+                                        // Middle of target is [...rest, ...T] and source is tuple type
+                                        // if T is constrained by a fixed-size tuple we might be able to use its arity to infer T
+                                        const param = getInferenceInfoForType(elementTypes[startLength + 1])?.typeParameter;
+                                        const constraint = param && getBaseConstraintOfType(param);
+                                        if (constraint && isTupleType(constraint) && !constraint.target.hasRestElement) {
+                                            const impliedArity = constraint.target.fixedLength;
+                                            const endIndex = sourceArity - getEndElementCount(target.target, ElementFlags.Fixed);
+                                            const startIndex = endIndex - impliedArity;
+                                            const trailingSlice = createTupleType(getTypeArguments(source).slice(startIndex, endIndex), source.target.elementFlags.slice(startIndex, endIndex),
+                                                /*readonly*/ false, source.target.labeledElementDeclarations && source.target.labeledElementDeclarations.slice(startIndex, endIndex));

```

Comment:
This is based on the content of the `sliceTupleType`. However, I wasn't able to reuse that function as the purpose of this slice is different here and `sliceTupleType` wouldn't work for it - I didn't find another instance of such a "trailing slice" in the codebase so I've just inlined the thing here and adjusted for the needs here.

Reply:
To give more context behind this - I can't reuse `sliceTupleType` as it creates array types when `index > target.fixedLength`:
https://github.dev/microsoft/TypeScript/blob/acf854b636e0b8e5a12c3f9951d4edfa0fa73bcd/src/compiler/checker.ts#L15758-L15764

So it's not a literal "slice" - it's not just slicing between the `startIndex` and `endIndex` - it also performs a type conversion conditionally. 

If not that condition, I could reuse the `sliceTupleType` here as follows:

```ts
const impliedArity = constraint.target.fixedLength;
const targetEndLength = getEndElementCount(target.target, ElementFlags.Fixed);

inferFromTypes(sliceTupleType(source, sourceArity - targetEndLength - impliedArity, targetEndLength), elementTypes[startLength + 1]);
```